types/src/util/tag.rs

```diff
@@ -0,0 +1,68 @@
+// Copyright 2018 Kodebox, Inc.
+// This file is part of CodeChain.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use super::super::transaction::HashingError;
+
+pub struct Tag {
+    pub input_scheme: bool,
+    pub output_scheme: bool,
+    pub filter_len: usize,
+    pub filter: Vec<u8>,
+}
+
+impl Tag {
+    pub fn new(mut bitvec: Vec<u8>) -> Result<Tag, HashingError> {
+        let tag = if let Some(tag) = bitvec.pop() {
+            tag
+        } else {
+            return Err(HashingError::InvalidFilter)
+        };
+        let input_scheme = (tag & 0x1) == 1;
+        let output_scheme = (tag >> 1 & 0x1) == 1;
+        let filter_len = ((tag >> 2) & 0x3f) as usize;
+
+        if bitvec.len() != filter_len {
+            return Err(HashingError::InvalidFilter)
+        }
+
+        Ok(Tag {
+            input_scheme,
+            output_scheme,
+            filter_len,
+            filter: bitvec,
+        })
+    }
+
+    pub fn get_tag(self) -> Vec<u8> {
```

Comment:
How about making `bitvec` as a member and return the reference to that?

Reply:
I changed it.